VPLAY-10563 curl timeout disambiguation

lashmintha opened the pull request:
Reason for change: Distinguished the curl error 28, into DNS timeout error, connection with the host timeout and data transfer timeout
Risks: Low
Test Procedure: Refer jira ticket
Priority: P1

LashminthaJ committed:
VPLAY-10563 curl timeout disambiguation

Reason for change: Distinguished the curl error 28, into DNS timeout error, connection with the host timeout and data transfer timeout
Risks: Low
Test Procedure: Refer jira ticket
Priority: P1

Signed-off-by: lashmintha <lashminthaj17@gmail.com>

Vinish100 commented on AampMPDDownloader.cpp:
```diff
@@ -468,8 +468,31 @@ void AampMPDDownloader::downloadMPDThread1()
+			
 			if(mMPDData->mMPDDownloadResponse->iHttpRetValue != 200 && mMPDData->mMPDDownloadResponse->iHttpRetValue != 204 && mMPDData->mMPDDownloadResponse->iHttpRetValue != 206)
 			{ 
+				double nameLookupTime = 0;
```
This should be checked when the curl returns 28 or 18.. If we get an httpRetValue it means we got a response from server and didn't fail in connect

Vinish100 commented on AampMPDDownloader.cpp:
```diff
@@ -468,8 +468,31 @@ void AampMPDDownloader::downloadMPDThread1()
+				else if (connectTime != 0 && nameLookupTime != 0)
+				{
+					// Data transfer timeout (connection established)
+					AAMPLOG_WARN("Timeout: Data transfer failed for %s", tuneUrl.c_str());
```
These info has to be bubbled up to application through anomaly event and mediafailed event

Vinish100 commented on priv_aamp.cpp:
```diff
@@ -4289,12 +4289,37 @@ bool PrivateInstanceAAMP::GetFile( std::string remoteUrl, AampMediaType mediaTyp
+						else if (connectTime != 0 && nameLookupTime != 0)
+						{
+							// Data transfer timeout (connection established)
+							AAMPLOG_WARN("Timeout: Data transfer failed for %s", remoteUrl.c_str());
```
Same goes here..

Vinish100 commented on priv_aamp.cpp:
```diff
@@ -4289,12 +4289,37 @@ bool PrivateInstanceAAMP::GetFile( std::string remoteUrl, AampMediaType mediaTyp
+						if (connectTime == 0 && nameLookupTime == 0)
+						{
+							// DNS timeout
+							AAMPLOG_WARN("Timeout: DNS resolution failed for %s", remoteUrl.c_str());
```
Also is it possible to integrate these logs to LogNetworkError, to avoid logging a failed download multiple times?

LashminthaJ committed:
Revert "VPLAY-10777: Index 0 does not maps to lowest profile (#457)"

This reverts commit 23698d000958023dfafb6d1a8457b6f8e84e3f47.

LashminthaJ committed:
VPLAY-10563 curl timeout disambiguation

Reason for change: Distinguished the curl error 28, into DNS timeout error, connection with the host timeout and data transfer timeout
Risks: Low
Test Procedure: Refer jira ticket
Priority: P1

Signed-off-by: lashmintha <lashminthaj17@gmail.com>

Vinish100 commented on AampLogManager.h:
```diff
@@ -231,8 +231,16 @@ class AampLogManager
-					logprintf( eLOGLEVEL_ERROR, __FUNCTION__, __LINE__, "AAMPLogNetworkError error='curl error %d' type='%s' location='%s' symptom='%s' url='%s'",
+					if( failureReason != nullptr )
+					{
+						logprintf( eLOGLEVEL_ERROR, __FUNCTION__, __LINE__, "AAMPLogNetworkError error='curl error %d' type='%s' reason='%s' location='%s' symptom='%s' url='%s'",
```
This should be under AAMPNetworkErrorTimeout isn't it?
error='curl error %d(failureReason)' would be apt I guess, so that we don't add new fields for this additional info

Vinish100 commented on AampMPDDownloader.cpp:
```diff
@@ -468,6 +468,36 @@ void AampMPDDownloader::downloadMPDThread1()
 
 			mMPDData->mMPDStatus	=	AAMPStatusType::eAAMPSTATUS_MANIFEST_DOWNLOAD_ERROR;
+			
+			if( mMPDData->mMPDDownloadResponse->iHttpRetValue == CURLE_OPERATION_TIMEDOUT || mMPDData->mMPDDownloadResponse->iHttpRetValue == CURLE_COULDNT_RESOLVE_HOST || mMPDData->mMPDDownloadResponse->iHttpRetValue == CURLE_COULDNT_CONNECT )
```
Why CURLE_COULDNT_RESOLVE_HOST? Disambiguation is only required for timeout. I think I mentioned its required for partial file previously, sorry for misdirection

Vinish100 commented on AampMPDDownloader.cpp:
```diff
@@ -468,6 +468,36 @@ void AampMPDDownloader::downloadMPDThread1()
+				}
+				AampLogManager::LogNetworkError (mEffectiveUrl.c_str(), AAMPNetworkErrorCurl, mMPDData->mMPDDownloadResponse->iHttpRetValue, eMEDIATYPE_MANIFEST,mMPDData->mMPDDownloadResponse->failureReasonString);
+				//Use DownloadResponse Show call instead of printheaderresponse fn -since it is not scope
+				mMPDData->mMPDDownloadResponse->show();
```
Required?

Vinish100 commented on AampMPDDownloader.cpp:
```diff
@@ -468,6 +468,36 @@ void AampMPDDownloader::downloadMPDThread1()
+					// Data transfer timeout (connection established)
+					mMPDData->mMPDDownloadResponse->failureReasonString = "Data transfer failed";
+				}
+				AampLogManager::LogNetworkError (mEffectiveUrl.c_str(), AAMPNetworkErrorCurl, mMPDData->mMPDDownloadResponse->iHttpRetValue, eMEDIATYPE_MANIFEST,mMPDData->mMPDDownloadResponse->failureReasonString);
```
AAMPNetworkErrorTimeout

Vinish100 commented on fragmentcollector_mpd.cpp:
```diff
@@ -4586,7 +4588,7 @@ AAMPStatusType StreamAbstractionAAMP_MPD::FetchDashManifest()
 				aamp->UpdateDuration(0);
 				aamp->SetFlushFdsNeededInCurlStore(true);
-				aamp->SendDownloadErrorEvent(AAMP_TUNE_MANIFEST_REQ_FAILED, http_error);
+				aamp->SendDownloadErrorEvent(AAMP_TUNE_MANIFEST_REQ_FAILED, http_error, failureReason);
```
Isn't it better to define new AAMPTuneFailure for the failureReasons and use them rather than sending a new argument

Vinish100 commented on priv_aamp.cpp:
```diff
@@ -4288,12 +4295,38 @@ bool PrivateInstanceAAMP::GetFile( std::string remoteUrl, AampMediaType mediaTyp
+						//CURL* curl = mDownloader1.GetCurlHandle();
+						curl_easy_getinfo(curl, CURLINFO_NAMELOOKUP_TIME, &nameLookupTime);
+						curl_easy_getinfo(curl, CURLINFO_CONNECT_TIME, &connectTime);
+						if (connectTime == 0 && nameLookupTime == 0)
```
It would be better to define this logic in a common utility function, so that we can prevent any disparity from arising in future.

Vinish100 commented on priv_aamp.cpp:
```diff
@@ -4288,12 +4295,38 @@ bool PrivateInstanceAAMP::GetFile( std::string remoteUrl, AampMediaType mediaTyp
 						}
+						
+						//Curl error to be differentiated
+						AAMPLOG_INFO("Connection/Timeout error %d for url %s", res, remoteUrl.c_str());
```
Log required?

Vinish100 commented on priv_aamp.cpp:
```diff
@@ -4288,12 +4295,38 @@ bool PrivateInstanceAAMP::GetFile( std::string remoteUrl, AampMediaType mediaTyp
+						AAMPLOG_INFO("Connection/Timeout error %d for url %s", res, remoteUrl.c_str());
+						//Curl error to be differentiated
+						double nameLookupTime = 0;
+    					double connectTime = 0;
```
Only needed for timeout

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -4586,7 +4588,31 @@ AAMPStatusType StreamAbstractionAAMP_MPD::FetchDashManifest()
+					{	
+						aamp->SendDownloadErrorEvent(AAMP_TUNE_DNS_CACHE_TIMEOUT, http_error);
+					}
+					else if ( failureReason == "Connection to host failed" )
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
03ba04bd152ff000fb157939e1a943c9
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Pointer comparison with string literal
Comparing string literal ""Connection to host failed"" to "failureReason" via direct pointer comparison is unlikely to have the desired effect.

_Medium Impact, CWE-597_
_BAD_COMPARE_

## How to fix
 Did you intend to use something like strcmp() or wcscmp() instead?

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -4586,7 +4588,31 @@ AAMPStatusType StreamAbstractionAAMP_MPD::FetchDashManifest()
+					{
+						aamp->SendDownloadErrorEvent(AAMP_TUNE_CURL_CONNECTION_TIMEOUT, http_error);
+					}
+					else if( failureReason == "Data transfer failed" )
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
26064351afda9c74e87bf49ecd58da21
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Pointer comparison with string literal
Comparing string literal ""Data transfer failed"" to "failureReason" via direct pointer comparison is unlikely to have the desired effect.

_Medium Impact, CWE-597_
_BAD_COMPARE_

## How to fix
 Did you intend to use something like strcmp() or wcscmp() instead?

</details>

rdkcmf-jenkins commented on priv_aamp.cpp:
```diff
@@ -2638,7 +2638,22 @@ void PrivateInstanceAAMP::SendDownloadErrorEvent(AAMPTuneFailure tuneFailure, in
-						snprintf(description,MAX_DESCRIPTION_SIZE, "%s : Curl Error Code %d, Download time expired", tuneFailureMap[tuneFailure].description, error_code);
+						if( tuneFailure == AAMP_TUNE_DNS_CACHE_TIMEOUT )
+						{
+							snprintf(description,MAX_DESCRIPTION_SIZE, "%s : Curl Error Code %d, DNS resolution failed", tuneFailureMap[tuneFailure].description, error_code);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
0eae36fe803712e63f3a51d847a4e83f
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Logically dead code
Execution cannot reach this statement: "snprintf(description, 128UL...".

_Medium Impact, CWE-561_
_DEADCODE_



</details>

rdkcmf-jenkins commented on priv_aamp.cpp:
```diff
@@ -2638,7 +2638,22 @@ void PrivateInstanceAAMP::SendDownloadErrorEvent(AAMPTuneFailure tuneFailure, in
+						}
+						else if( tuneFailure == AAMP_TUNE_CURL_CONNECTION_TIMEOUT )
+						{
+							snprintf(description,MAX_DESCRIPTION_SIZE, "%s : Curl Error Code %d, Curl connection to host failed", tuneFailureMap[tuneFailure].description, error_code);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
0eae36fe803712e63f3a51d847a4e83f
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Logically dead code
Execution cannot reach this statement: "snprintf(description, 128UL...".

_Medium Impact, CWE-561_
_DEADCODE_



</details>

rdkcmf-jenkins commented on priv_aamp.cpp:
```diff
@@ -2638,7 +2638,22 @@ void PrivateInstanceAAMP::SendDownloadErrorEvent(AAMPTuneFailure tuneFailure, in
+						}
+						else if( tuneFailure == AAMP_TUNE_DATA_TRANSFER_FAILURE )
+						{
+							snprintf(description,MAX_DESCRIPTION_SIZE, "%s : Curl Error Code %d, Data transfer failed", tuneFailureMap[tuneFailure].description, error_code);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
0eae36fe803712e63f3a51d847a4e83f
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Logically dead code
Execution cannot reach this statement: "snprintf(description, 128UL...".

_Medium Impact, CWE-561_
_DEADCODE_



</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -4586,7 +4588,31 @@ AAMPStatusType StreamAbstractionAAMP_MPD::FetchDashManifest()
+
+				if( http_error == CURLE_OPERATION_TIMEDOUT && failureReason != nullptr )
+				{
+					if( failureReason == "DNS resolution failed")
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
1c6320f46d19f2bc459c73bfdf3772be
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Pointer comparison with string literal
Comparing string literal ""DNS resolution failed"" to "failureReason" via direct pointer comparison is unlikely to have the desired effect.

_Medium Impact, CWE-597_
_BAD_COMPARE_

## How to fix
 Did you intend to use something like strcmp() or wcscmp() instead?

</details>

rdkcmf-jenkins commented on priv_aamp.cpp:
```diff
@@ -4288,12 +4303,39 @@ bool PrivateInstanceAAMP::GetFile( std::string remoteUrl, AampMediaType mediaTyp
+						{
+							failureReason = GetCurlTimeoutFailureReason(curl);
+							//Curl error to be differentiated
+							if (failureReason == "DNS resolution failed")
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
c340f22661dd4e639a98d37ab3567e9e
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Pointer comparison with string literal
Comparing string literal ""DNS resolution failed"" to "failureReason" via direct pointer comparison is unlikely to have the desired effect.

_Medium Impact, CWE-597_
_BAD_COMPARE_

## How to fix
 Did you intend to use something like strcmp() or wcscmp() instead?

</details>

rdkcmf-jenkins commented on priv_aamp.cpp:
```diff
@@ -4288,12 +4303,39 @@ bool PrivateInstanceAAMP::GetFile( std::string remoteUrl, AampMediaType mediaTyp
+								// Connection timeout (DNS succeeded)
+								SendDownloadErrorEvent(AAMP_TUNE_CURL_CONNECTION_TIMEOUT, CURLE_OPERATION_TIMEDOUT);
+							}
+							else if (failureReason == "Data transfer failed")
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
246ab8b5a684350f64a579cdd5165d30
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Pointer comparison with string literal
Comparing string literal ""Data transfer failed"" to "failureReason" via direct pointer comparison is unlikely to have the desired effect.

_Medium Impact, CWE-597_
_BAD_COMPARE_

## How to fix
 Did you intend to use something like strcmp() or wcscmp() instead?

</details>

Vinish100 commented on AampMPDDownloader.cpp:
```diff
@@ -682,6 +690,22 @@ ManifestDownloadResponsePtr AampMPDDownloader::GetManifest(bool bWait, int iWait
+				// Optionally, log or use the handle
+				if (curlHandle)
+				{
+					respPtr->mMPDDownloadResponse->failureReasonString = GetCurlTimeoutFailureReason(curlHandle);
```
Why is this required? Here the download is still ongoing, we are not stopping it. But we are checking the curl info values. If working add a detailed comment about this block

Vinish100 commented on AampUtils.cpp:
```diff
@@ -36,7 +36,7 @@
 #include <ctime>
 #include <cctype>
-#include <curl/curl.h>
+//#include <curl/curl.h>
```
Remove if not required

Vinish100 commented on priv_aamp.cpp:
```diff
@@ -213,6 +213,9 @@ static TuneFailureMap tuneFailureMap[] =
 	{AAMP_TUNE_INVALID_MANIFEST_FAILURE, 10, "AAMP: Invalid Manifest, parse failed"},
 	{AAMP_TUNE_FAILED_PTS_ERROR, 80, "AAMP: Playback failed due to PTS error"},
 	{AAMP_TUNE_MP4_INIT_FRAGMENT_MISSING, 10, "AAMP: init fragments missing in playlist"},
+	{AAMP_TUNE_DNS_CACHE_TIMEOUT, 60, "AAMP: DNS resolution failed due to timeout"},
```
Why 60? It should be 10 as the playback didn't start

Vinish100 replied:
Manifest download failed due to DNS resolve timeout

Vinish100 commented on priv_aamp.cpp:
```diff
@@ -2638,7 +2641,22 @@ void PrivateInstanceAAMP::SendDownloadErrorEvent(AAMPTuneFailure tuneFailure, in
 						break;
 				case CURLE_OPERATION_TIMEDOUT:
-						snprintf(description,MAX_DESCRIPTION_SIZE, "%s : Curl Error Code %d, Download time expired", tuneFailureMap[tuneFailure].description, error_code);
+						if( tuneFailure == AAMP_TUNE_DNS_CACHE_TIMEOUT )
```
Why special cases here. Already tuneFailure can be used to print the appropriate description

Vinish100 commented on AampMPDDownloader.cpp:
```diff
@@ -468,6 +468,14 @@ void AampMPDDownloader::downloadMPDThread1()
+			
+			if( mMPDData->mMPDDownloadResponse->iHttpRetValue == CURLE_OPERATION_TIMEDOUT  )
+			{ 
+				mMPDData->mMPDDownloadResponse->failureReasonString = GetCurlTimeoutFailureReason(mDownloader1.GetCurlHandle());
```
Call inside below if

Vinish100 commented on fragmentcollector_mpd.cpp:
```diff
@@ -4586,7 +4588,31 @@ AAMPStatusType StreamAbstractionAAMP_MPD::FetchDashManifest()
+				{
+					if( strcmp(failureReason,"DNS resolution failed")==0)
+					{	
+						aamp->SendDownloadErrorEvent(AAMP_TUNE_DNS_CACHE_TIMEOUT, http_error);
```
AAMP_TUNE_DNS_RESOLVE_TIMEOUT

Vinish100 commented on fragmentcollector_mpd.cpp:
```diff
@@ -4586,7 +4588,31 @@ AAMPStatusType StreamAbstractionAAMP_MPD::FetchDashManifest()
+					}
+					else if( strcmp(failureReason,"Data transfer failed" )==0 )
+					{
+						aamp->SendDownloadErrorEvent(AAMP_TUNE_DATA_TRANSFER_FAILURE, http_error);
```
AAMP_TUNE_DATA_TRANSFER_TIMEOUT

Vinish100 commented on streamabstraction.cpp:
```diff
@@ -2386,7 +2386,7 @@ void StreamAbstractionAAMP::GetDesiredProfileOnBuffer(int currProfileIndex, int
 	{
 		AAMPLOG_WARN("Switch to index 0; buffer is about to drain :Buffer %lf !!",bufferValue);
-		newProfileIndex = aamp->mhAbrManager.getProfileIndexForLowestBandwidth();
+		newProfileIndex = 0;
```
Rebase your code on top of latest sprint
